File: js/dll.js
'use strict'

class Node{
  constructor(val, next = null, prev = null){
    this.val = val
    this.next = next
    this.prev = prev
  }
}

class DLL{
  constructor(){
    this.head = null
    this.tail = null
    this._counter = 0
  }

  push(val){
    let oldHead = this.head
    this.head = new Node(val, this.next = this.head)
    this._counter ++
    if(oldHead){
      oldHead.prev = this.head
    }else{
      this.tail = this.head
    }
  }

  append(val){
    let oldTail = this.tail
    this.tail = new Node(val, this.prev = this.tail)
    this._counter ++
    if(oldTail){
      oldTail.next = this.tail
    }else{
      this.head = this.tail
    }
  }

  pop(){
    if(this._counter === 0){
      this.head = null
      this.tail = null
    }
    if(!this.head){
      throw new Error('Node not in list.')
    }
    let output = this.head.val
    this.head = this.head.next
    if(this.head){
      this.head.prev = null
    }else{
      this.tail = null
    }
    this._counter --
    return output
  }

  shift(){
    if(this._counter === 0){
      this.head = null
      this.tail = null
    }
    if(!this.tail){
      throw new Error('Node not in list.')
    }
    let output = this.tail.val
    this.tail = this.tail.prev
    if(this.tail){
      this.tail.next = null
    }else{
      this.head = null
    }
    this._counter --
    return output
  }

  remove(val){
    if(!this.head){
      return 'List is empty.'
    }
    if(this.head.val === val){
      this.pop()
      return
    }
    let curr = this.head
    while(curr.next){
      if(curr.val === val){
        curr.prev.next = curr.next
        curr.next.prev = curr.prev
        this._counter --
        return
      }
      curr = curr.next
    }
    if(val === this.tail.val){
      this.shift()
      return
    }
    return 'Node not in list.'
  }
}

module.exports = DLL

File: src/dll.py
"""Implementation of double linked list."""

from linked_list import LinkedList


class Node(object):
    """Class for node."""

    def __init__(self, val, next_node=None, prev_node=None, priority=0):
        """Create a new node."""
        self.val = val
        self.next_node = next_node
        self.prev_node = prev_node


class DLL(LinkedList):
    """Double linked list."""

    def __init__(self):
        """On list creation set tail to none."""
        super(DLL, self).__init__()
        self.tail = None

    def push(self, val):
        """Add new item to the head of the double linked list."""
        if self.head is None:
            self.head = Node(val, self.head)
            self.tail = self.head
        else:
            orig_head = self.head
            self.head = Node(val, self.head)
            orig_head.prev_node = self.head
        self.length += 1

    def append(self, val, priority=0):
        """Add new item to the tail of the double linked list."""
        if self.tail is None:
            self.tail = Node(val, priority=priority)
            self.head = self.tail
        else:
            new_node = Node(val, priority=priority)
            orig_tail = self.tail
            self.tail = new_node
            self.tail.prev_node = orig_tail
            orig_tail.next_node = self.tail
        self.length += 1

    def pop(self):
        """Remove and return the head."""
        if self.head is None:
            raise IndexError('List is empty, cannot pop from an empty list')
        val = self.head.val
        if self.length == 1:
            self.head = None
            self.tail = None

        else:
            self.head = self.head.next_node
            self.head.prev_node = None

        self.length -= 1
        return val

    def shift(self):
        """Remove and return the tail."""
        if self.head is None:
            raise IndexError('List is empty, cannot pop from an empty list')
        val = self.tail.val
        if self.length == 1:
            self.head = None
            self.tail = None
        else:
            self.tail = self.tail.prev_node
            self.tail.next_node = None

        self.length -= 1
        return val

    def remove(self, val):
        """Remove selected node."""
        if not self.head:
            raise IndexError('List is empty, cannot pop from an empty list')
        current = self.head
        while current:
            if current.val == val:
                if current.prev_node:
                    current.prev_node.next_node = current.next_node
                    current.next_node.prev_node = current.prev_node
                else:
                    self.head = current.next_node
                    current.next_node.prev_node = None
            current = current.next_node
        self.length -= 1
